hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationPublishersContainer.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.wan.impl;
+
+import com.hazelcast.internal.services.ServiceNamespace;
+import com.hazelcast.monitor.LocalWanPublisherStats;
+import com.hazelcast.spi.partition.PartitionReplicationEvent;
+import com.hazelcast.wan.WanReplicationEvent;
+import com.hazelcast.wan.WanReplicationPublisher;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.hazelcast.util.MapUtil.createHashMap;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+/**
+ * Delegating WAN replication publisher implementation. This implementation
+ * is a container for multiple WAN publishers.
+ * When publishing an event on this delegate, all publishers are notified.
+ */
+public final class WanReplicationPublishersContainer {
+    /** Non-null WAN replication name */
+    final String name;
+    /** Non-null WAN publishers, grouped by publisher ID */
+    final ConcurrentMap<String, WanReplicationPublisher> publishers;
+
+    public WanReplicationPublishersContainer(@Nonnull String name,
+                                             @Nonnull ConcurrentMap<String, WanReplicationPublisher> publishers) {
+        checkNotNull(name, "WAN publisher name should not be null");
+        checkNotNull(publishers, "WAN publisher map should not be null");
+        this.name = name;
+        this.publishers = publishers;
+    }
+
+    /** Returns all {@link WanReplicationPublisher}s for this delegate */
+    public @Nonnull
+    Collection<WanReplicationPublisher> getPublishers() {
+        return publishers.values();
+    }
+
+    /**
+     * Returns the {@link WanReplicationPublisher} with the {@code publisherId}
+     * or {@code null} if it doesn't exist.
+     */
+    public WanReplicationPublisher getPublisher(String publisherId) {
+        return publishers.get(publisherId);
+    }
+
+    public void addPublisher(@Nonnull String publisherId,
```

Comment:
OS is not using this method. Can we move this kind of unused methods to EE?

Reply:
Again, I'm not too concerned with it as it's private and moving it elsewhere might create other kind of complexity. One idea is that we might have an OS `WanReplicationPublishersContainer` with some methods and then an EE `EnterpriseWanReplicationPublishersContainer` extending the OS one. But one point of this PR was to eliminate the OS-EE distinction. WDYT?

Reply:
i think this is a minor point, no action is ok.

Reply:
I'll leave this for now as I'd like to focus on getting as much breaking changes done until feature freeze. You are right about it, though and I'll add it to the list of todo's in the comment above.